internal/app/orchestrator/orchestrator.go

```diff
@@ -95,16 +115,197 @@ func New(ctx context.Context,
+	err = o.cache.AddRequest(aggregatedKey, &req)
+	if err != nil {
+		// If we fail to register the watch, we need to kill this stream by
+		// closing the response channel.
+		o.logger.With("err", err).With("key", aggregatedKey).With(
+			"req node", req.GetNode()).Error(ctx, "failed to add watch")
+		closedChannel := o.downstreamResponseMap.delete(&req)
+		return closedChannel, nil
+	}
+
+	// Check if we have a cached response first.
+	cached, err := o.cache.Fetch(aggregatedKey)
+	if err != nil {
+		// Log, and continue to propagate the response upstream.
+		o.logger.With("err", err).With("key", aggregatedKey).Warn(ctx, "failed to fetch aggregated key")
+	}
+
+	if cached != nil && cached.Resp != nil && cached.Resp.Raw.GetVersionInfo() != req.GetVersionInfo() {
+		// If we have a cached response and the version is different,
+		// immediately push the result to the response channel.
+		go func() { responseChannel <- convertToGcpResponse(cached.Resp, req) }()
+	}
+
+	// Check if we have a upstream stream open for this aggregated key. If not,
+	// open a stream with the representative request.
+	if !o.upstreamResponseMap.exists(aggregatedKey) {
+		upstreamResponseChan, shutdown, err := o.upstreamClient.OpenStream(req)
+		if err != nil {
+			// TODO implement retry/back-off logic on error scenario.
+			// https://github.com/envoyproxy/xds-relay/issues/68
+			o.logger.With("err", err).With("key", aggregatedKey).Error(ctx, "Failed to open stream to origin server")
+		} else {
+			respChannel, upstreamOpenedPreviously := o.upstreamResponseMap.add(aggregatedKey, upstreamResponseChan)
+			if upstreamOpenedPreviously {
+				// A stream was opened previously due to a race between
+				// concurrent downstreams for the same aggregated key, between
+				// exists and add operations. In this event, simply close the
+				// slower stream and return the existing one.
+				shutdown()
+			} else {
+				// Spin up a go routine to watch for upstream responses.
+				// One routine is opened per aggregate key.
+				go o.watchUpstream(ctx, aggregatedKey, respChannel.response, respChannel.done, shutdown)
+			}
+		}
+	}
+
+	return responseChannel, o.onCancelWatch(aggregatedKey, &req)
 }
 
 // Fetch implements the polling method of the config cache using a non-empty request.
-func (c *orchestrator) Fetch(context.Context, discovery.DiscoveryRequest) (*gcp.Response, error) {
+func (o *orchestrator) Fetch(context.Context, discovery.DiscoveryRequest) (*gcp.Response, error) {
 	return nil, fmt.Errorf("Not implemented")
 }
 
-func (c *orchestrator) onCacheEvicted(key string, resource cache.Resource) {
-	// TODO implement.
+// watchResponse is intended to be called in a goroutine, to receive incoming
+// responses and fan out to downstream clients.
+func (o *orchestrator) watchUpstream(
+	ctx context.Context,
+	aggregatedKey string,
+	responseChannel <-chan *discovery.DiscoveryResponse,
+	done <-chan bool,
+	shutdownUpstream func(),
+) {
+	for {
```

Comment:
This logic really hurts my head. It would be great to add a comment here about how it listens for responses from the upstream channel, caches the response  and fans out to the change to the downstream channel.

It would also be good to document how this shuts down and the purpose of the shutdownUpstream function. I assume it is via the done channel.

Reply:
oh no :( 

I do document the first part at the [function header](https://github.com/envoyproxy/xds-relay/pull/53/files#diff-2958638164db92742469d2bb08d0001aR192-R193) and make comments throughout, although the documentation for done/shutdownUpstream could be more clear. I'll put more details in the function header comment.